src/Prelude.purs

```diff
@@ -811,13 +786,43 @@ instance ordChar :: Ord Char where
 instance ordUnit :: Ord Unit where
   compare _ _ = EQ
 
-instance ordArray :: (Ord a) => Ord [a] where
-  compare [] [] = EQ
-  compare [] _ = LT
-  compare _ [] = GT
-  compare (x:xs) (y:ys) = case compare x y of
-    EQ -> compare xs ys
-    other -> other
+instance ordArray :: (Ord a) => Ord (Array a) where
+  compare = compareArrayImpl checkEqOr EQ
+    where
+    checkEqOr a b callback =
+      case compare a b of
+        EQ -> unit
+        other -> callback other
+
+foreign import compareArrayImpl
+  """
+  function compareArrayImpl(checkEqOr) {
+    return function(eq) {
+      return function(xs) {
+        return function(ys) {
+          var minlen = xs.length
+          var maxlen = ys.length
+          var result = null
+
+          if (maxlen < minlen) {
+             var tmp = minlen
+             minlen = maxlen
+             maxlen = tmp
+          }
+
+          for (var i = 0; i < minlen && result === null; ++i) {
+            checkEqOr(xs[i])(ys[i])(function(other) { result = other })
+          }
+
+          if (result === null) {
+             result = eq
+           }
+          return result
+        }
+      }
+    }
+  }
+  """ :: forall a. (a -> a -> (Ordering -> Unit) -> Unit) -> Ordering -> Array a -> Array a -> Ordering
```

Comment:
See the version in `tests/prelude.purs` in the compiler repo. It uses `Number`, but you get the gist.